apm-agent-core/src/main/java/co/elastic/apm/agent/bci/HelperClassManager.java

```diff
@@ -128,52 +167,169 @@ public T getForClassLoaderOfClass(Class<?> classOfTargetClassLoader) {
             return localHelper;
         }
 
-        private static <T> T createHelper(@Nullable ClassLoader targetClassLoader, ElasticApmTracer tracer, String implementation, String... additionalHelpers) {
-            try {
-                final Map<String, byte[]> typeDefinitions = getTypeDefinitions(asList(implementation, additionalHelpers));
-                Class<? extends T> helperClass;
-                try {
-                    helperClass = loadHelperClass(targetClassLoader, implementation, typeDefinitions);
-                } catch (ClassNotFoundException | NoClassDefFoundError e) {
-                    // in the unit tests, the agent is not added to the bootstrap class loader
-                    helperClass = loadHelperClass(ClassLoader.getSystemClassLoader(), implementation, typeDefinitions);
+        @Override
+        protected void markFailedClassLoader(ClassLoader failedClassLoader) {
+            failed = true;
+        }
+    }
+
+    public static class ForAnyClassLoader<T> extends HelperClassManager<T> {
+
+        // doesn't need to be concurrent - invoked only from a synchronized context
+        static final Map<Integer, WeakReference<List<Object>>> clId2helperImplListMap = new HashMap<>();
+
+        final ConcurrentHashMap<Integer, WeakReference<T>> clId2helperMap;
```

Comment:
Why don't you use WeakConcurrentMaps here?

Reply:
Yes, I did, but it seems to be more costly in the overall

Reply:
I'd be surprised if `WeakConcurrentMap`s are slower. What is it you are concerned about? The weak references for the key? The lookup should be just as fast and even allocation free. It would also simplify `loadAndReferenceHelper` as it takes care of cleaning the stale entries.

Reply:
I thought of using `WeakReference`s as keys, but these maps shouldn't grow a lot as they have entry-per-class-loader. Changed to `WeakConcurrentMap`